cassandra/cluster.py

```diff
@@ -1734,12 +1734,16 @@ def protocol_downgrade(self, host_endpoint, previous_version):
         self.protocol_version = new_version
 
     def _add_resolved_hosts(self):
+        endpoints = list(h[1].endpoint for h in self.metadata.all_hosts_items())
```

Comment:
Maybe a more robust solution would be to move that check into `self.add_host` - more specifically `self.metadata.add_or_return_host` which acquires `_hosts_lock`. And if you look at the implementation of that function:

https://github.com/sylwiaszunejko/python-driver/blob/a1e14f17e3b7e5251d092426ab06e76273a84b7f/cassandra/metadata.py#L328-L340

it does a very similar thing - it tries to avoid adding a duplicate `Host`, but it only looks at `host_id`, not `endpoint`.

Reply:
Actually the code in `add_or_return_host` used to look at `endpoint` and not `host_id`, but this commit changed that... :(

https://github.com/scylladb/python-driver/commit/8b78f068c63c2a85fc3a0b081a516969cdde73ff#diff-74dd2f135e6a6c0a300c937d529cc2b93fa5f7faf47b1c94fc4dea0005476cddL331-L337

Reply:
@avelanarius there is a problem with adding the check in `add_or_return_host`, this will affect not only `add_host` which is used in `_add_resolved_hosts`, but also `update_host` which is used to update when we read from system.peers, in this case we want to add host with the same endpoint. Maybe adding a check in `add_host` - with acquiring a lock of course - would be sufficient?

Reply:
@avelanarius I pushed that version with check in `add_host` for you to look if it is acceptable

Reply:
Just to be sure, I assume that you re-did the testing described here (https://github.com/scylladb/python-driver/pull/304#issuecomment-1971495589)?

Reply:
>  but also update_host which is used to update when we read from system.peers, in this case we want to add host with the same endpoint.

What's the case that you are referring to? (I'm not sure if I'm thinking about the same case). 

Also I think you should describe it in more detail (and if it makes sense add this description to code/commit message).

Reply:
> Just to be sure, I assume that you re-did the testing described here ([#304 (comment)](https://github.com/scylladb/python-driver/pull/304#issuecomment-1971495589)) with that version?

yes

Reply:
> > but also update_host which is used to update when we read from system.peers, in this case we want to add host with the same endpoint.
> 
> What's the case that you are referring to? (I'm not sure if I'm thinking about the same case).
> 
> Also I think you should describe it in more detail (and if it makes sense add this description to code/commit message).

I am referring to the logic in `_refresh_node_list_and_token_map` - https://github.com/scylladb/python-driver/blob/8138dca6427176a4eccb6b872769b16e31769ad7/cassandra/cluster.py#L4004 here we run `update_host` and this function adds new host based on system.peers, after running `update_host` there is a logic in `_refresh_node_list_and_token_map`  to remove hosts that are not present in system.peers.